### PR TITLE
Convert description and description-dom-formatter to typeScript

### PR DESCRIPTION
**Pull request by shans**

### CODE REVIEW

File: runtime/ts/description.ts [hunk 1]
```diff
@@ -176,15 +198,19 @@ export class DescriptionFormatter {
...
         return jointDescription;
       } else {
         return this._capitalizeAndPunctuate(jointDescription);
       }
     }
+    return undefined;
   }
 
-  _joinDescriptions(strings) {
+  // TODO(mmandlis): the override of this function in subclasses also overrides the output. We'll need to unify
```
Comment (mmandlis) [inline runtime/ts/description.ts]: hmm, interesting - this one returns a string, but the dom formatter returns an object... i guess we need a DescriptionOutput class then?

File: runtime/ts/description.ts [hunk 2]
```diff
@@ -9,34 +9,41 @@
...
 export class Description {
+  relevance: {};
+  readonly arc: Arc;
+  _particle: Particle | undefined = undefined;
+
   constructor(arc) {
-    this._arc = arc;
-    this._relevance = null;
+    this.arc = arc;
+    this.relevance = null;
```
[the remark this answers is not in the document]
Comment (shans) [reply runtime/ts/description.ts]: The statement above says that relevance is of type object, it doesn't set a value to it. I can combine the two though.

File: runtime/ts/description-dom-formatter.ts [hunk 3]
```diff
@@ -9,13 +9,14 @@
...
-import {DescriptionFormatter} from './description.js';
+import {assert} from '../../platform/assert-web.js';
+import {DescriptionFormatter, CombinedDescriptionsOptions, ParticleDescription} from './description.js';
 
 export class DescriptionDomFormatter extends DescriptionFormatter {
+  private nextID = 0;
   constructor(description) {
     super(description);
-    this._nextID = 0;
+    this.nextID = 0;
```
[the remark this answers is not in the document]
Comment (shans) [reply runtime/ts/description-dom-formatter.ts]: Done! This means we don't need an extra constructor at this level, too :)

File: runtime/ts/description.ts (hunk 1 above)
Comment (shans) [reply runtime/ts/description.ts]: There are other alternatives too:
(1) list the possible outputs as a union (but this means you need to know the possible types in the superclass)
(2) separate the subclass functionality out as not using the same method names.

**Pull request merged**